packages/lib-webpack/src/webpack/DynamicRemotePlugin.ts

```diff
@@ -0,0 +1,167 @@
+    };
+    return acc;
+  }, {} as WebpackSharedModules);
+
+/**
+ * Minimal subset of webpack `SharedConfig` type.
+ *
+ * @see https://webpack.js.org/plugins/module-federation-plugin/#sharing-hints
+ */
+type WebpackSharedModuleConfig = {
+  import?: string | false;
+  singleton?: boolean;
+};
+
+/**
+ * Equivalent to webpack `SharedObject` type.
+ */
+type WebpackSharedModules = { [moduleRequest: string]: WebpackSharedModuleConfig };
+
+export type DynamicRemotePluginOptions = Partial<{
+  /**
+   * Plugin metadata JSON file name, or the parsed plugin metadata.
+   *
+   * Default value: `plugin.json`.
+   */
+  pluginMetadata: string | PluginBuildMetadata;
+
+  /**
+   * List of extensions contributed by the plugin.
+   *
+   * The value is either a `minimatch` compatible JSON file glob pattern,
+   * or the parsed extensions array.
+   *
+   * Default value: `extensions.json`.
+   */
+  extensions: string | Extension[];
+
+  /**
+   * Modules shared between the host application and its plugins at runtime.
+   *
+   * Essential modules like `react` and `redux` will be added automatically.
+   *
+   * Default value: empty object.
+   */
+  sharedModules: WebpackSharedModules;
+}>;
+
+export class DynamicRemotePlugin implements webpack.WebpackPluginInstance {
+  private readonly pluginMetadata: PluginBuildMetadata;
+
+  private readonly extensions: Extension[];
+
+  private readonly sharedModules: WebpackSharedModules;
+
+  constructor(options: DynamicRemotePluginOptions = {}) {
+    const adaptedOptions: Required<DynamicRemotePluginOptions> = {
+      pluginMetadata: options.pluginMetadata ?? 'plugin.json',
+      extensions: options.extensions ?? 'extensions.json',
+      sharedModules: options.sharedModules ?? {},
+    };
+
+    // Load and validate plugin metadata
+    if (typeof adaptedOptions.pluginMetadata === 'string') {
+      this.pluginMetadata = loadPluginMetadata(adaptedOptions.pluginMetadata);
+    } else {
+      this.pluginMetadata = validatePluginMetadata(adaptedOptions.pluginMetadata);
+    }
+
+    // Load and validate extensions contributed by the plugin
+    if (typeof adaptedOptions.extensions === 'string') {
+      this.extensions = loadExtensions(adaptedOptions.extensions);
+    } else {
+      this.extensions = validateExtensions(adaptedOptions.extensions);
+    }
+
+    // Create plugin vs. host application shared module configuration
+    this.sharedModules = {
+      ...adaptedOptions.sharedModules,
+      ...getSharedModulesWithStrictSingletonConfig(['react', 'redux']),
+    };
+  }
+
+  apply(compiler: webpack.Compiler) {
+    const logger = compiler.getInfrastructureLogger(DynamicRemotePlugin.name);
+    const containerName = this.pluginMetadata.name;
+
+    if (!compiler.options.output.publicPath) {
+      logger.warn(
+        'output.publicPath is not defined, which may cause plugin assets to not load properly in the browser',
+      );
+    }
+
+    if (!compiler.options.output.uniqueName) {
```

Comment:
@andrewballantyne If a plugin explicitly declares `output.uniqueName` then we assume it has the right value.

What we could do here is compare it to `pluginMetadata.name` and fail if the values aren't equal.

Reply:
Not sure why we would allow them to specify if the only value they can give us is equal to the `name`? Wouldn't we just set it to that anyways?

Reply:
@andrewballantyne Technically, `output.uniqueName` doesn't have to equal `pluginMetadata.name` for the plugin to work at runtime. `output.uniqueName` is used for scoping module chunks containing the given plugin's exposed modules, which are loaded by host application at runtime.

We cannot prevent plugins to not specify webpack `output.uniqueName` options :smile: if we want to have it strictly equal to `pluginMetadata.name` then we can fail the build early.

Reply:
That's fair haha... I was thinking more why would we allow them to ever use it... flat out be annoying and throw an error if we can only have one value and it's already set somewhere else 😛

Reply:
Agreed, will update code to fail if `output.uniqueName` is specified explicitly, since this is something our webpack plugin will set automatically.